File: src/2025-02-22-00-51-50--a-deepseek-r1-test-220-10-1--2-1-huoshan/javascript/exercises/practice/variable-length-quantity/variable-length-quantity.js
//
// This is only a SKELETON file for the 'Variable Length Quantity' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const encode = (numbers) => {
  return numbers.flatMap(number => {
    if (number === 0) return [0x00];
    
    const chunks = [];
    while (number > 0) {
      chunks.push(number & 0x7f);
      number = number >>> 7;
    }
    
    chunks.reverse();
    
    for (let i = 0; i < chunks.length - 1; i++) {
      chunks[i] |= 0x80;
    }
    
    return chunks;
  });
};

export const decode = (bytes) => {
  const result = [];
  let currentValue = 0;
  let continuation = false;

  for (const byte of bytes) {
    if (byte & 0x80) {
      continuation = true;
    } else {
      continuation = false;
    }

    currentValue = (currentValue << 7) | (byte & 0x7f);

    if (!continuation) {
      result.push(currentValue >>> 0); // Convert to unsigned 32-bit
      currentValue = 0;
    } else if (byte === bytes[bytes.length - 1]) {
      throw new Error('Incomplete sequence');
    }
  }

  if (continuation) {
    throw new Error('Incomplete sequence');
  }

  return result;
};

File: src/2025-02-22-00-51-50--a-deepseek-r1-test-220-10-1--2-1-huoshan/go/exercises/practice/variable-length-quantity/variable_length_quantity.go
package variablelengthquantity

import "errors"

func EncodeVarint(input []uint32) []byte {
	var result []byte
	for _, n := range input {
		if n == 0 {
			result = append(result, 0x00)
			continue
		}

		var chunks []byte
		for n > 0 {
			chunks = append(chunks, byte(n&0x7F))
			n >>= 7
		}

		// Reverse chunks
		for i, j := 0, len(chunks)-1; i < j; i, j = i+1, j-1 {
			chunks[i], chunks[j] = chunks[j], chunks[i]
		}

		// Set high bits on all but last chunk
		for i := 0; i < len(chunks)-1; i++ {
			chunks[i] |= 0x80
		}

		result = append(result, chunks...)
	}
	return result
}

func DecodeVarint(input []byte) ([]uint32, error) {
	var result []uint32
	var current uint64

	for _, b := range input {
		current = (current << 7) | uint64(b&0x7F)

		if current > 0xFFFFFFFF {
			return nil, errors.New("overflow")
		}

		if (b & 0x80) == 0 {
			result = append(result, uint32(current))
			current = 0
		} else if current > (0xFFFFFFFF >> 7) {
			return nil, errors.New("overflow")
		}
	}

	if current != 0 {
		return nil, errors.New("incomplete sequence")
	}

	return result, nil
}
